flink-core/src/main/java/org/apache/flink/api/common/operators/ResourceSpec.java

```diff
@@ -183,17 +238,81 @@ public int hashCode() {
 		result = 31 * result + directMemoryInMB;
 		result = 31 * result + nativeMemoryInMB;
 		result = 31 * result + stateSizeInMB;
+		result = 31 * result + extendedResources.hashCode();
 		return result;
 	}
 
 	@Override
 	public String toString() {
+		String extend = "";
+		for (Resource resource : extendedResources.values()) {
+			extend += ", " + resource.name + "=" + resource.value;
+		}
 		return "ResourceSpec{" +
 				"cpuCores=" + cpuCores +
 				", heapMemoryInMB=" + heapMemoryInMB +
 				", directMemoryInMB=" + directMemoryInMB +
 				", nativeMemoryInMB=" + nativeMemoryInMB +
-				", stateSizeInMB=" + stateSizeInMB +
+				", stateSizeInMB=" + stateSizeInMB + extend +
 				'}';
 	}
+
+	private void addResource(String name, double value, ResourceAggregateType type) {
+		extendedResources.put(name, new Resource(name, type, value));
+	}
+
+	public static class Resource {
+		private String name;
+		private ResourceAggregateType type;
+		private Double value;
+
+		public Resource(String name, double value) {
+			this(name, ResourceAggregateType.AGGREGATE_TYPE_SUM, value);
```

Comment:
`ResourceAggregateType` should be the last argument since `name` and `value` are properly passed to this method.